packages/boot/README.md

```diff
@@ -0,0 +1,109 @@
+# @loopback/boot
+
+Boot package for LoopBack 4 to bootstrap convention based projects.
+
+## Overview
+
+This package provides the ability to bootstrap a LoopBack 4 project by
+automatically automatically associating artifacts and configuration with an
+application at runtime.
+
+The package is currently consumed as a [Mixin](http://loopback.io/doc/en/lb4/Mixin.html).
+
+It will automatically find all [Controller](http://loopback.io/doc/en/lb4/Controllers.html)
+classes by searching through all files in `controllers` directory ending in 
+`.controller.js` and bind them to the application using `this.controller()`.
+
+Other Mixins can support automatic Booting by overriding the `boot()` method.
+See example in [Advanced use](#advanced-use).
+
+## Installation
+
+```sh
+npm i --save @loopback/boot
+```
+
+## Basic use
+
+Using `@loopback/boot` is simple. It is a Mixin and should be added to your
+application Class as shown below.
+
+```ts
+class BootedApplication extends BootMixin(Application) {
+  constructor(options?: ApplicationConfig) {
+    super(options);
+  }
+}
+```
+
+### Configuration
+
+Configuration options for boot can be set via the `ApplicationConfig` object 
+(`options`). The following options are supported:
+
+|Name|Default|Description|
+|-|-|-|
+|`boot.rootDir`|`process.cwd()`|The root directory for the project. All other paths are resolved relative to this path.|
+|`boot.controllerDirs`|`controllers`|String or Array of directory paths to search in for controllers. Paths may be relative to `rootDir` or absolute.|
+|`boot.controllerExts`|`controller.js`|String or Array of file extensions to consider as controller files.|
+
+*Example*
+```ts
+app = new BootedApplication({
+  rest: {
+    port: 3000
+  },
+  boot: {
+    rootDir: '/absolute/path/to/my/app/root',
+    controllerDirs: ['controllers', 'ctrls'],
+    controllerExts: ['controller.js', '.ctrl.js'],
+  },
+});
+
+## Advanced use
+
+An extension developer can support the **boot** process by writing a mixin and
+overriding the `boot()` function as shown below. If a user is using
+`@loopback/boot`, it will automatically boot your artifact as well.
```

Comment:
Again, I don't think `Mixin` is the right approach here. It has the same issue as Express `app.use` as the order of execution is controlled by the order how mixins are chained. I would rather use the `extension point/extension` pattern as we do in LoopBack 3.x.

Please note that a single `boot()` method is not going to work.  There are multiple steps needed to `discover/load/transform/resolve/start/stop` various artifacts which can be cross-referenced. We need to allow each extensions to participate the boot process at different phases.

I strongly recommend that we follow a similar approach at https://github.com/strongloop/loopback-boot/tree/master.

Please

Reply:
Thanks for the comments Raymond. Looking for some clarification. 

1. My understanding of `loopback-boot` right now is that it runs scripts found in a `boot` folder alphabetically / in a given order by requiring them. We can accomplish the same using the Mixin by running scripts in a boot folder. This task was just about discovering and binding artifacts. 

2. I sort of see the value of having different functions for `discover/load/transform/etc.` but if the purpose of boot is to find and load artifacts than I can imagine a single function being able to do that as a extension developer can write the boot function to do what they want. 

3. I'm not sure I understand the purpose of certain phases that are proposed such as `resolve/start/stop`. Resolve in particular is something that I think is better handled by Dependency Injection.

With @kjdelisle 's change, you can control Mixin order by passing them in via an array. 

Just trying to understand the value add of having phases for boot vs. just booting by discovering artifacts via Mixin followed by running scripts in a boot folder like current `loopback-boot`.

Reply:
> My understanding of loopback-boot right now is that it runs scripts found in a boot folder alphabetically / in a given order by requiring them. We can accomplish the same using the Mixin by running scripts in a boot folder. This task was just about discovering and binding artifacts.

No. Boot script is just one of the artifact types we support in LoopBack 3.x. There are different plugins for model definitions, data sources, middleware, components, and boot scripts. Each plugin is responsible for loading/compiling/executing its own kind of artifact.

> I sort of see the value of having different functions for discover/load/transform/etc. but if the purpose of boot is to find and load artifacts than I can imagine a single function being able to do that as a extension developer can write the boot function to do what they want.

One pass is usually not good enough to handle artifacts that have references to other ones. For example, model `Customer` has a `hasMany` relation to model `Order` and `Order` has a `belongsTo` relation to `Customer`. A single `boot()` for models cannot build the runtime representation of the relations between the two models .

> I'm not sure I understand the purpose of certain phases that are proposed such as resolve/start/stop. Resolve in particular is something that I think is better handled by Dependency Injection.

Resolve - connect multiple artifacts, for example, set up the target model reference for a relation.
Start - start an artifact, for example, a server instance or a database connection
Stop - stop an artifact, for example, a server instance or a database connection

Reply:
I see. But shouldn't references be resolved by Dependency Injection when needed if boot sets up bindings properly which is what it should be doing ...?  

And I see the `mixins` folder on `loopback-boot`. So it controls loading of artifacts by calling the start() for each mixin?

I'm also not sure what it means to start / stop an artifact ... or how stop will even be called?

Reply:
> I see. But shouldn't references be resolved by Dependency Injection when needed if boot sets up bindings properly which is what it should be doing ...?

DI resolution is different from the metadata dereferencing. We use `@inject` to describe dependencies for `services`, but not for metadata.

> And I see the mixins folder on loopback-boot. So it controls loading of artifacts by calling the start() for each mixin?

Where do you see it? https://github.com/strongloop/loopback-boot/tree/master/lib/plugins?

> I'm also not sure what it means to start / stop an artifact ... or how stop will even be called?

`start` and `stop` can go beyond `boot`. We might have `app.start` and `app.stop` to use these methods.

Reply:
Yea sorry I meant `plugins` not mixins.

Reply:
@raymondfeng Your proposed changes are out of scope for the task. If this is something we want, then we should create a new task, groom it and poker it.

Reply:
IMO, the approach taken by the PR is architecturally flawed. I'm not sure if we really want to land it as is. The boot module is fundamental to the declarative support for LoopBack 4. I think we need to start with the right path from the beginning.

I don't mind starting with a single `boot()` implemented by individual booters, but `mixin` is not suitable for this use case.

Reply:
Great discussion! 

Let me clarify few possible confusions I anticipate from the discussion above before I post my opinions.

In LoopBack 3.x, there are two versions of loopback-boot. `2.x` is what almost everybody uses today, because it's the latest version published to npmjs.org and used by our CLI tooling when scaffolding a new project. It has rather simple and less flexible design, the order of individual boot steps is hard coded and there are no extension points for components. Then there is the upcoming new version 3.0 living in the master branch containing a rewrite by @raymondfeng and @davidcheung, which provides much more flexible design with extension points for components. I believed this version was never published (see https://github.com/strongloop/loopback-boot/issues/241), but as I checked the master branch today, there are apparently some 3.x versions published to npmjs.org. Because the release process of the new semver-major version was not done properly, these 3.x versions are effectively hidden by 2.x updates. What a mess! But let's not digress - I just wanted to clarify the state of loopback-boot to prevent confusion created by people (e.g. from our community) reading the discussion above.

> My understanding of loopback-boot right now is that it runs scripts found in a boot folder alphabetically / in a given order by requiring them. We can accomplish the same using the Mixin by running scripts in a boot folder. This task was just about discovering and binding artifacts.

This is definitely the case in loopback-boot 2.x. I don't know how the new 3.x version handles boot scripts.

> And I see the mixins folder on loopback-boot. So it controls loading of artifacts by calling the start() for each mixin?

Even though this has been clarified that @virkt25 meant `plugins, not `mixins`, I'd like to note that in LoopBack 3.x, we have a concept of [Model Mixins](https://loopback.io/doc/en/lb3/Defining-mixins.html) that allow users to apply a mixin to a model class, and at least loopback-boot 2.x does load and execute such mixin scripts.


---

> [@kjdelisle] @raymondfeng Your proposed changes are out of scope for the task. If this is something we want, then we should create a new task, groom it and poker it.
> [@raymondfeng] IMO, the approach taken by the PR is architecturally flawed. I'm not sure if we really want to land it as is. The boot module is fundamental to the declarative support for LoopBack 4. I think we need to start with the right path from the beginning.
> I don't mind starting with a single boot() implemented by individual booters, but mixin is not suitable for this use case.

I agree with both of you.

IMO, we should not aim for a fully flexible bootstrapper in this first iteration targeting our MVP milestone. For MVP, it's ok to have a hard-coded bootstrapper with no (or very little) extension points.

At the same time, I agree that mixin composition is not a good architecture for our bootstrapper and therefore we should not pursue it even for MVP, because upgrading from mixin composition to a different architecture would be too disruptive.

TBH, I find Raymond's `BootLifeCycle` with pre-defined steps as suboptimal too. Assuming each of the steps like `discover` and `load` are run fully for all components before the next step is executed (i.e. run `discover` for all components and only then run `load` for all components), then how are we going to handle the situation when ComponentA needs the output of `load` step from ComponentB before it can `discover` its own artifacts? On the other hand, if entire `BootLifeCycle` is run for each component before the next component is booted (i.e. run `discover` and `load` for ComponentA, then for ComponentB), then how are we going to handle the situation when ComponentA needs to run some task after ComponentB has loaded & transformed all its artifacts, but before ComponentB attempts to resolve them?

I am proposing to explore our "Sequence" design pattern we are using for composing request handling middleware and see if and how it can be applied to the boot process too. I see two benefits of such approach:

 - By reusing the same concept for both request handling logic and bootstrapping, users can apply their existing knowledge in multiple places. The framework should be easier to learn as a result.

 - Sequence makes the flow control explicit, easy to understand and modify.

Right now, the boot process (both `master` (3.x) and `2.x` branches of loopback-boot for LoopBack 3.x) is very opaque. The current 2.x branch does not allow any customization of boot steps - see the hard-coded algorithm in [executor.js](https://github.com/strongloop/loopback-boot/blob/5060032d265bd457c14ea2510ecd0d017a7a2f1b/lib/executor.js#L33-L42). The new version in `master` (which hasn't been released for public use yet - see https://github.com/strongloop/loopback-boot/issues/241) seems to promise more flexibility in that matter, but my knowledge of that new codebase is not sufficient to tell how well we are delivering on that promise.

Let me post a mock-up boot sequence to illustrate my idea:

```ts
type EnvironmentVariables = {[name: string]: string};

interface BootOptions {
  projectRoot: string;
  env?: EnvironmentVariables;
}

interface AppBootSequence<App> {
  boot(app: App, options: BootOptions);
}

class MyBootSequence implements AppBootSequence<MyApplication> {
   /* constructor receiving dependencies - I am leaving that out for brevity */

  public boot(app: MyApplication, options: BootOptions) {
    this.setupDataSources(app, options);
    this.setupRepositoriesAndModels(app, options);
    this.resolveModelRelations(app, options);
    this.setupControllers(app, options);
    this.start(app, env);
  }
}

// setupDataSources step provided by @loopback/repository or similar
setupDataSources(app, options) {
  const controllersDir = path.resolve(
    options.projectRoot,
    'controllers' || options.controllersDir);

  const controllerExts = options.controllerExts || 'controller.js';

  bootClassArtifacts(app, controllersDir, controllerExts, a => app.controller(a));
}

// usage in Application
// this code should be scaffolded by our CLI tooling
// together with MyBootSequence template
class MyApplication {
  constructor(/*...*/) {
    this.bind('bootstrapper').toClass(MyBootSequence)
  }

  async boot(env: EnvironmentVariables) {
    const options: BootOptions = {
      projectRoot: __dirname,
      env: env,
    }
    const sequence = await this.get('bootstrapper');
    await sequence.boot(app, options);
  }
}
```

Since this design is similar with REST Sequence Handler, I would expect it should be reasonably easy to leverage the extension point https://github.com/strongloop/loopback-next/pull/657 and action composition https://github.com/strongloop/loopback-next/pull/713 features that @raymondfeng is working on, to get an even more flexible solution.

Thoughts?

Reply:
Let me expand a bit on why I am against mixin composition.

Consider `ComponentA` and `ComponentB`, where `ComponentA` needs to run it's boot function only after `ComponentB` has already booted. This requires `ComponentB` to call `super.boot()` before running its own boot steps and the application developer to compose mixins in the correct order. As a result, neither the ComponentB nor the application developer are fully in control of the order in which individual boot steps are invoked and the correct order of boot steps requires cooperation of multiple parts of the codebase. I see that as a sign of Low [Cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) & High [Coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming)), which typically makes code more difficult to understand and maintain.

(I admit my solution proposed above does not address this problem entirely, because in the Sequence design, individual sequence steps have no control of the order in which they are executed. OTOH, the action composition technique proposed by Raymond should address that part in mostly backwards-compatible way.)